### ISSUE TITLE
Import of ESM fails when declaration file present/not named as *.d.mts

### ISSUE THREAD

**Issue by blayet** [microsoft/TypeScript#50032]
# Bug Report

The Typescript compiler reports following error when importing an ESM package (using "type": "module"):
"error TS1471: Module 'minimal-lib/dist/math.js' cannot be imported using this construct. The specifier only resolves to an ES module, which cannot be imported synchronously. Use dynamic import instead."

The problem seems to be how the .d.ts is interpreted (i.e. as CJS not ESM).
Removing the .d.ts or renaming to .d.mts resolves error.

Note: I briefly tried using .mjs as lib file extension, but that seemed to have problems too. (Even though declaration was then named .d.mts.)

### 🔎 Search Terms

.d.ts

### 🕗 Version & Regression Information

Using 4.7.4.
Haven't tried in other versions.

### ⏯ Playground Link

sorry, not provided.

### 💻 Code

See minimal reproduction code:
https://github.com/blayet/minimal-ts-bug

### 🙁 Actual behavior

Error compiling a lib that contains typescript-generated declaration file.

### 🙂 Expected behavior

Compilation by tsc of a lib generated by tsc should work without having to rename files.

**Comment by IllusionMH** [comment]
Looks like duplicate of #50009

**Comment by blayet** [comment]
Not sure - that ticket talks about "Consider adding `"type": "module"` to this 'package.json' file."
That's already present in this case.

**Comment by IllusionMH** [comment]
It's present in `minimal-lib`, but `minimal-app` doesn't have `"type": "module"` and therefore `index.ts` would be treated as CommonJS.

From common JS you can't `require` ESM file (which is pointed by error) as `import` would've been transpired to `require` if `module` was `node16` (probably this is expectation for `"moduleResolution: "node16"`). If you add `"type": "module"` to `minimal-app` as error suggests - it's fixed.

With `"module": "es2020"` you have invalid CommonJS output as `import` isn't transpiled to `require`.

So not sure what is your expected runtime and use case for mixing `"module": "es2020"` and `"moduleResolution": "node16"`, but I think there is mismatch between `module`, `moduleResolution`, and `type` in configs for `minimal-app`

**Comment by blayet** [comment]
Huh. Yeah looks like you're right. (I thought I had "type": "module" in the app too, to be honest.)
Looks like I was confused by error as described in #50009 - as you say.
Thanks :)

Regarding runtime and use case. Currently its just for internal usage. I want to build a lib for use in web apps. I'll use webpack to bundle the app. Not sure about others downstream although we can probably assume/mandate use of a bundler - and definitely for web apps.
I picked "module": "es2020" because (reading https://www.typescriptlang.org/tsconfig#module) I noticed that "ES2020 adds support for dynamic imports, and import.meta." I thought that might be relevant  - not now but at some future point.
As far as I understand it any ESM choice would have been fine.
